File: map_ts/src/Exercices/Restaurant/03-product-availability.ts
import type {
  Ingredient,
  StockItem,
  PreparedProduct
} from "../../Restaurant/index.js";


const ingredients: Ingredient[] = [
  {
    id: "beef",
    name: "Beef",
    quantity: 3600,
    unit: "g",
    caloriesPerUnit: 2.5
  },
  {
    id: "chicken",
    name: "Chicken",
    quantity: 3600,
    unit: "g",
    caloriesPerUnit: 1.65
  },
  {
    id: "fish",
    name: "Fish",
    quantity: 3600,
    unit: "g",
    caloriesPerUnit: 1.2
  },
  {
    id: "potato",
    name: "Potato",
    quantity: 6000,
    unit: "g",
    caloriesPerUnit: 0.77
  },
  {
    id: "tomato",
    name: "Tomato",
    quantity: 6000,
    unit: "g",
    caloriesPerUnit: 0.18
  },
  {
    id: "salad",
    name: "Salad",
    quantity: 6000,
    unit: "g",
    caloriesPerUnit: 0.18
  },
  {
    id: "oignon",
    name: "Oignon",
    quantity: 6000,
    unit: "g",
    caloriesPerUnit: 0.32
  },
];


const stocks: StockItem[] = ingredients.map(
  ingredient => ({
    ingredient,
    availableQuantity: ingredient.quantity,
    state: "available"
  })
);


const products: PreparedProduct[] = [
  {
    id: "beef-plate",
    name: "Beef Plate",

    ingredients: [
      {
        ingredientId: "beef",
        quantity: 180
      },
      {
        ingredientId: "potato",
        quantity: 150
      }
    ],

    availableUnits: 0
  },
  {
    id: "chicken-plate",
    name: "Chicken Plate",

    ingredients: [
      {
        ingredientId: "chicken",
        quantity: 180
      },
      {
        ingredientId: "salad",
        quantity: 150
      }
    ],

    availableUnits: 0
  },
  {
    id: "fish-plate",
    name: "Fish Plate",

    ingredients: [
      {
        ingredientId: "fish",
        quantity: 180
      },
      {
        ingredientId: "oignon",
        quantity: 150
      }
    ],

    availableUnits: 0
  }

];


const availableProducts =
  products.map(product => {

    const possibleQuantities =
      product.ingredients.map(component => {

        // EXERCICE :
        //
        // 1. chercher le StockItem correspondant
        //    avec stocks.find(...)
        const stock = stocks.find(stock => stock.ingredient.id === component.ingredientId);
        // 2. calculer :
        //
        // stock disponible
        // ----------------
        // quantité requise
        const possibleQuantity = stock ? stock.availableQuantity / component.quantity : 0;
        // 3. retourner Math.floor(...)

        return Math.floor(possibleQuantity);
      });


    // EXERCICE :
    //
    // availableUnits doit être
    // la plus petite valeur
    // de possibleQuantities.

    return {
      ...product,

      availableUnits: Math.min(...possibleQuantities)
    };
  });


console.log(availableProducts);

File: map_python/src/Exercises/Restaurant/03_product_availability.py
from dataclasses import replace

from src.Restaurant import (
    Ingredient,
    PreparedProduct,
    ProductIngredient,
    StockItem,
    StockState,
)


ingredients = [
    Ingredient("beef", "Beef", 3600, "g", 2.5),
    Ingredient("chicken", "Chicken", 3600, "g", 1.65),
    Ingredient("fish", "Fish", 3600, "g", 1.2),
    Ingredient("potato", "Potato", 6000, "g", 0.77),
    Ingredient("salad", "Salad", 6000, "g", 0.18),
    Ingredient("oignon", "Oignon", 6000, "g", 0.32),
]

stocks = [
    StockItem(item, item.quantity, StockState.AVAILABLE)
    for item in ingredients
]

products = [
    PreparedProduct(
        "beef-plate",
        "Beef Plate",
        [ProductIngredient("beef", 180), ProductIngredient("potato", 150)],
    ),
    PreparedProduct(
        "chicken-plate",
        "Chicken Plate",
        [ProductIngredient("chicken", 180), ProductIngredient("salad", 150)],
    ),
    PreparedProduct(
        "fish-plate",
        "Fish Plate",
        [ProductIngredient("fish", 180), ProductIngredient("oignon", 150)],
    ),
]


def available_units(product: PreparedProduct) -> int:
    possible_quantities = list(
        map(
            lambda component: next(
                (
                    int(stock.available_quantity // component.quantity)
                    for stock in stocks
                    if stock.ingredient.id == component.ingredient_id
                ),
                0,
            ),
            product.ingredients,
        )
    )
    return min(possible_quantities, default=0)


available_products = list(
    map(
        lambda product: replace(product, available_units=available_units(product)),
        products,
    )
)


if __name__ == "__main__":
    for product in available_products:
        print(product)
